File: EPI/15.10/swift/15.10_test.swift
import Darwin

extension Node: CustomStringConvertible {
    public var description: String {
        return describe()
    }

    private func describe(indentation: String = "") -> String {
        let childIndentation = indentation + "  "

        let l = left?.describe(indentation: childIndentation) ?? "nil"
        let r = right?.describe(indentation: childIndentation) ?? "nil"

        return "Node(\n" +
            "\(childIndentation)value: \(self.value),\n" +
            "\(childIndentation)left: \(l),\n" +
            "\(childIndentation)right: \(r)\n" +
            "\(indentation))"
    }
}

func == (lhs: Node, rhs: Node) -> Bool {
    return lhs.value == rhs.value &&
        (lhs.left == nil && rhs.left == nil || lhs.left! == rhs.left!) &&
        (lhs.right == nil && rhs.right == nil || lhs.right! == rhs.right!)
}

let tests: [(sortedValues: [Int], tree: Node)] = [
    (
        sortedValues: [1, 2, 3],
        tree: Node(
            value: 2,
            left: Node(value: 1),
            right: Node(value: 3)
        )
    ),
    (
        sortedValues: [2, 3, 5, 7, 11, 13, 17, 19, 23],
        tree: Node(
            value: 11,
            left: Node(
                value: 3,
                left: Node(value: 2),
                right: Node(
                    value: 5,
                    left: nil,
                    right: Node(value: 7)
                )
            ),
            right: Node(
                value: 17,
                left: Node(value: 13),
                right: Node(
                    value: 19,
                    left: nil,
                    right: Node(value: 23)
                )
            )
        )
    ),
]

for test in tests {
    let actual = Node(sortedValues: test.sortedValues)
    guard actual == test.tree else {
        print("For sorted values \(test.sortedValues), " +
            "expected minimum-height ST to be \(test.tree), but was \(actual)")
        exit(1)
    }
}

File: EPI/15.10/typescript/15.10_test.ts
import {Node, createBalancedBST} from './15.10';

declare function require(name: string): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

const trimUndefined = (tree: Node | undefined) => {
    if (!tree) {
        return undefined;
    }

    const copy: Node = {value: tree.value};

    if (tree.left) {
        copy.left = trimUndefined(tree.left);
    }

    if (tree.right) {
        copy.right = trimUndefined(tree.right);
    }

    return copy;
}

const tests: {values: number[], tree: Node | undefined}[] = [
    {
        values: [],
        tree: undefined
    },
    {
        values: [1, 2, 3],
        tree: {
            value: 2,
            left: {value: 1},
            right: {value: 3}
        }
    },
    {
        values: [2, 3, 5, 7, 11, 13, 17, 19, 23],
        tree: {
            value: 11,
            left: {
                value: 5,
                left: {
                    value: 3,
                    left: {value: 2}
                },
                right: {value: 7}
            },
            right: {
                value: 19,
                left: {
                    value: 17,
                    left: {value: 13}
                },
                right: {value: 23}
            }
        }
    },
];

for (const test of tests) {
    const actual = trimUndefined(createBalancedBST(test.values));
    assert.deepStrictEqual(actual, test.tree, `For values ${test.values}, ` +
        `expected tree to be ${inspect(test.tree)}, but was ${inspect(actual)}`);
}
